File: Java/src/Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> ret = new ArrayList<>(); 
        List<Integer> stack = new ArrayList<>(); 
        if (root == null) return ret;
        
        pathStack(root, sum, stack, ret);
        return ret;
        
    }
        
    private void pathStack(TreeNode root, int sum, List<Integer> stack, List<List<Integer>> ret){
        stack.add(root.val);
        sum -= root.val;
        
        if(root.right != null){
            pathStack(root.right, sum, new ArrayList<>(stack), ret);
        }
        
        if(root.left != null){
            pathStack(root.left, sum, new ArrayList<>(stack), ret);
        }
        
        if(root.left == null && root.right == null && sum == 0){
            ret.add(stack);
        }
    }
}

File: Python/Path Sum II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, summary):
        if not root:
            return []
        ret = []
        stack = []
        self.pathstack(root, summary, stack, ret)
        return ret
    
    def pathstack(self, root, summary, stack, ret):
        stack.append(root.val)
        summary = summary - root.val
        if root.right:
            self.pathstack(root.right, summary, stack[:], ret)
        if root.left:
            self.pathstack(root.left, summary, stack[:], ret)
        if root.left == None and root.right == None and summary == 0:
            return ret.append(stack)
        
        
    
        
        